tests/python/test_python_frontend.py

```diff
@@ -4055,6 +4055,93 @@ def fusion_func(fd: FusionDefinition) -> None:
                 "FusionDefinition's execute() did not run correctly with profile enabled!"
             )
 
+    # Test that the range of generated uniform values spans the proper range
+    # https://github.com/NVIDIA/Fuser/issues/1653
+    def test_uniform_range(self):
+        dtypes = [DataType.Double, DataType.Float, DataType.Half]
+        if not is_pre_ampere():
+            dtypes.append(DataType.BFloat16)
+        for dtype in dtypes:
+            samples_per_run = 2**29
+
+            def fusion_fn(fd: FusionDefinition):
+                # Generate enough values to reasonably expect to sample the ends of the range
+                shape = fd.define_vector([samples_per_run], dtype=DataType.Int)
+                S0 = fd.define_scalar(0.00000, dtype=DataType.Double)
+                S1 = fd.define_scalar(1.00000, dtype=DataType.Double)
+                output = fd.ops.uniform(S0, S1, shape=shape, dtype=dtype)
+                fd.add_output(output)
+
+            with FusionDefinition() as fd:
+                fusion_fn(fd)
+
+            output = fd.execute([])[0]
+
+            x = output.amax()
+            mu = output.type(torch.float64).mean()
+            # Repeat to improve chances of sampling extreme values
+            num_runs = 100
+            num_samples = num_runs * samples_per_run
+            for i in range(num_runs):
+                u = fd.execute([])[0]
+                x = torch.maximum(x, u.amax())
+                mu = mu + (u.type(torch.float64).mean() - mu) / (i + 2)
+
+            if dtype != DataType.Double:
+                # Even with repeats, there's no hope of sampling extreme double values
+
+                assert (
+                    x.item()
+                    == torch.nextafter(
+                        torch.tensor(1.0, dtype=output.dtype),
+                        torch.tensor(0.0, dtype=output.dtype),
+                    ).item()
```

Comment:
This check is achievable since the precision at 1.0 is much coarser than near zero. For types with large exponent like bfloat16, we will not generate samples at exactly 0.0. This differs from pytorch actually: they do generate from 0 to nextafter(1, 0) for each datatype I tested. We should probably have a look at how they're implementing `rand` for bfloat16.

Reply:
```python
import torch      
      
def run(dtype):      
    m = None      
    x = None      
    for _ in range(1000):      
        u = torch.rand(2**30, dtype=dtype, device='cuda')      
        m = u.amin() if m is None else torch.minimum(m, u.amin())      
        x = u.amax() if x is None else torch.maximum(x, u.amax())      
      
    theomax = torch.nextafter(      
            torch.tensor(1.0, dtype=dtype), torch.tensor(0.0, dtype=dtype))      
    print(f'{dtype} {m.item()} {x.item()}, theoretical max {theomax.item()}')      
      
run(torch.float64)      
run(torch.float32)      
run(torch.float16)      
run(torch.bfloat16) 
```
> torch.float64 3.615163723935666e-12 0.9999999999995994, theoretical max 0.9999999999999999
torch.float32 0.0 0.9999999403953552, theoretical max 0.9999999403953552
torch.float16 0.0 0.99951171875, theoretical max 0.99951171875
torch.bfloat16 0.0 0.99609375, theoretical max 0.99609375

Reply:
One way to go would be to generate using 64 bit ints instead of 32 bit, if using bf16. We could still use 32bit ints for float and fp16.